### ISSUE TITLE
Need to speed up spectral map creation / PCA

### ISSUE THREAD

**Issue by ksens** [Paradigm4/revealgenomics#115]
@arko-bhattacharya 

At deployment A, the following workflow is carried out:

- RNA-seq gene, biosample, expression value data.frame for a specific pipeline (a.k.a. MeasurementSet) is downloaded from SciDB, 
- the data.frame is converted into a 2D matrix (gene * biosample)
- the matrix is decorated (with relevant metadata) into a Bioconductor ExpressionSet
- `esetVis::createSpectralMap` function is called thrice to create plots w.r.t. the first three principal components. 

As you can see in timing snapshots below (run on demo machine), the spectral map creation piece is fast enough for small matrices, but can be prohibitively slow for larger matrices. 

```R
###################
# smallest RNA-seq matrix in TCGA: TCGA-CHOL

eset1 = search_expression(measurementset = ms1, biosample_ref = bios_ref1) 
# Downloading full dataset from SciDB ... done (4.579 sec)
# Downloading features to form ExpressionSet ... done (1.877 sec)
# Forming ExpressionSet ... Reshaping expr to Matrix ... Handling continuous dense matrix case ... done (0.157 sec)

system.time({
  xx = createSpectralMap(eSetObj = eset1)
})[3]
# elapsed 
#   8.248 


###################
# largest RNA-seq matrix in TCGA: TCGA-BRCA

bios_ref2 = search_biosamples(dataset_id = id2)
eset2 = search_expression(measurementset = ms2, biosample_ref = bios_ref2) 
# Downloading full dataset from SciDB ... done (44.654 sec)
# Downloading features to form ExpressionSet ... done (2.243 sec)
# Forming ExpressionSet ... Reshaping expr to Matrix ... Handling continuous dense matrix case ... done (6.027 sec)

system.time({
  xx = createSpectralMap(eSetObj = eset2)
})[3]
# elapsed 
# 956.989
```

Please look into different options of speeding up the spectral map creation. Two strategies:
1. Look at options within Bioconductor / R to speed up the PCA
2. Use the scidb PCA calculation operator for the PCA

## Notes

`0`. I will share the repro code for running spectral map creation on TCGA demo machine on this ticket soon. 

`1`. If you use a different option for running the PCA (which I assume is taking the longest time), you will need to do the necessary data munging to create the equivalent output for `createSpectralMap` (eventually, we want a clean drop-in into the deployment code) e.g. deployment A depends on at least the following two items from the result of `createSpectralMap`

```
x$spectral12$analysis$axesContributionsPercentages
x$spectral12$plot
```

`2`. For approach 2 (using  scidb for PCA), I will need to share scidb code with you. Till I get to that, take a look into what `esetVis::createSpectralMap` really does.

**Comment by ksens** [comment]
Repro code updated (Note `0` above).

**Comment by arko-bhattacharya** [comment]
createSpectralMap computes PCA thrice.

enhanced_createSpectralMap() function in the above comit computes PCA once and uses this PCA to plot principal components.

Here are the computation time for TCGA-CHOL and TCGA-BRCA using createSpectralMap and enhanced_createSpectralMap:

> dat = get_data(study = 'TCGA-CHOL')
> system.time({smap1 = createSpectralMap(dat$exp)})[3]
> elapsed
> 8.296
> system.time({smap5 = enhanced_createSpectralMap(dat$exp)})[3]
> elapsed
> 3.568
> all.equal(smap1, smap5)
> [1] TRUE

> dat = get_data(study = 'TCGA-BRCA')
> system.time({smap1 = createSpectralMap(dat$exp)})[3]
> elapsed
> 950.208
> system.time({smap5 = enhanced_createSpectralMap(dat$exp)})[3]
> elapsed
> 320.770
> all.equal(smap1, smap5)
> [1] TRUE

**Comment by arko-bhattacharya** [comment]
Upon investigating mpm() which is called in esetVis::esetSpectralMap() for computing SVD we can see that the steps which take the most time in mpm() are pre SVD ETL and computing SVD. Here is a plot which tries to identify the time taken for these steps on a few TCGA data-sets.
![pca_vs_prePcaEtl_runtime](https://user-images.githubusercontent.com/16749721/66161514-1709d700-e64a-11e9-8613-1041ebc6cb5c.png)

**Comment by arko-bhattacharya** [comment]
The above image helps us in understanding that the maximum time in mpm() is spent in computing SVD. Hence, to further optmize enhanced_createSpectralMap() mpm()'s SVD computation has to be optimized.

**Comment by arko-bhattacharya** [comment]
mpm() calls La.svd() from R-base to compute SVD. This can be further optimized by using fast.svd() from library 'corpcor' OR by computing truncated SVD using propack.svd() from library 'svd'. The following plot shows the compute time per these methods on a few TCGA data-sets.
![pca_runTime_Lasvd_fastsvd_propacksvd](https://user-images.githubusercontent.com/16749721/66162164-6f8da400-e64b-11e9-9b54-f516048a83ff.png)
Note: Truncated SVD computed the first 20 principal components.

The above plot clearly shows that computing a truncated SVD using propack.svd() is the most optimized way to compute SVD.

**Comment by arko-bhattacharya** [comment]
Updated enhanced_createSpectralMap() with an optimized SVD method. 

Here are the plots showing the run time of various createSpectralMap() (CSM) functions. The following plots contain customers original CSM, enhanced CSM (whereby SVD is conducted only once), enhanced CSM with fast.svd() and enhanced CSM with a trucated SVD using propack.svd().

Plot with all TCGA data-sets.
![createSpectralMap_pca_methods_runTime](https://user-images.githubusercontent.com/16749721/66234375-33ba1380-e70b-11e9-9b93-b38e01ef0ea6.png)

Plot with TCGA data-sets without BRCA.
![createSpectralMap_pca_methods_runTime_outlierTreated](https://user-images.githubusercontent.com/16749721/66234447-5fd59480-e70b-11e9-93da-7b5777abd9b8.png)

**Comment by ksens** [comment]
This is resolved with #138 .
